pkg/controller/resourceclaim/controller.go

```diff
@@ -854,11 +857,35 @@ func (ec *Controller) syncClaim(ctx context.Context, namespace, name string) err
 		// those, the resource claim controller will trigger deletion when the
 		// pod is done. However, it doesn't hurt to also trigger deallocation
 		// for such claims and not checking for them keeps this code simpler.
-		if len(valid) == 0 &&
-			claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer {
-			claim.Status.DeallocationRequested = true
+		if len(valid) == 0 {
+			if builtinControllerFinalizer >= 0 {
+				if claim.Spec.AllocationMode == resourcev1alpha2.AllocationModeWaitForFirstConsumer ||
+					claim.DeletionTimestamp != nil {
+					// Allocated by scheduler with structured parameters. We can "deallocate"
+					// by clearing the allocation.
+					claim.Status.Allocation = nil
+					claim.Finalizers = slices.Delete(claim.Finalizers, builtinControllerFinalizer, builtinControllerFinalizer+1)
```

Comment:
I don't think you can update finalizers via UpdateStatus.

Update:
I talked with @liggitt about this code: https://github.com/kubernetes/kubernetes/blob/e568a77a931a1cf4239a4a5fa43e2b05bad3abdf/pkg/registry/resource/resourceclaim/strategy.go#L124-L127
It should call ResetObjectMetaForStatus

Reply:
Is that a "should" or a "must"?

I don't have that because it's not done in other resources either that I based my work on, for example:
https://github.com/kubernetes/kubernetes/blob/b0ee3343747520db7e41c298d566026cd8df0be8/pkg/registry/core/persistentvolumeclaim/strategy.go#L156-L162

In this case, being able to add or remove the finalizer together with the status update is desirable. If it has to be done in two API calls, there is an intermediate, incomplete state (finalizer set, status not) which then also needs to be considered.

Reply:
It's a must.
Older APIs are buggy, but we shouldn't repeat the mistake.

Reply:
I filed https://github.com/kubernetes/kubernetes/issues/123727.

Can we keep the code as-is in this PR and I add the bug fix before 1.30 gets released? Just trying to figure out our options. Obviously merging code which depends on buggy behavior isn't ideal, but missing code freeze is (IMHO) worse.

Reply:
https://github.com/kubernetes/kubernetes/blame/b0ee3343747520db7e41c298d566026cd8df0be8/staging/src/k8s.io/apimachinery/pkg/apis/meta/v1/helpers.go#L262

says
> This is opt-in for new objects with Status subresource.

So that should read "This is required for new objects with Status subresource."?

Reply:
This intermediate state isn't that hard to handle. I'm testing a fix.

Reply:
See https://github.com/kubernetes/kubernetes/pull/123730 for the bug-fix only PR.

I've updated the scheduler and controller code to do two API calls. The scheduler continues when interrupted in the middle and the controller cleans up if needed, so I think this should be okay.

Reply:
And if the finalizer is added but the next update fails, the controller should be able to remove the finalizer, right? as the allocation is empty.

Could there be a race between scheduler and the controller? Maybe there should be a timeout before deleting the finalizer?

Reply:
That race was indeed also my concern. I'm solving it by only removing the finalizer of an unallocated claim if it also has a deletion time stamp. In that state it is unambiguous that the scheduler is not going to add the allocation in parallel.

Reply:
I see. In the past, I've used `map[string]interface{}` to represent the yaml before marshaling. Although it's unclear which one reads better.